pkg/karmadactl/promote/promote.go

```diff
@@ -788,3 +745,66 @@ func buildClusterPropagationPolicy(resourceName, policyName, cluster string, gvr
 	}
 	return cpp
 }
+
+// ensureResourceInControlPlaneWithOptionalPolicy ensures that a specified resource is present in the Karmada control plane.
+// It optionally creates propagation policies for the resource.
+func ensureResourceInControlPlaneWithOptionalPolicy(controlPlaneDynamicClient dynamic.Interface, karmadaClient karmadaclientset.Interface,
+	obj *unstructured.Unstructured, o *CommandPromoteOption, gvr schema.GroupVersionResource,
+	propagateDeps bool, isNamespaceScoped bool) error {
+	if isNamespaceScoped {
+		_, err := controlPlaneDynamicClient.Resource(gvr).Namespace(o.Namespace).Get(context.TODO(), o.name, metav1.GetOptions{})
+		if err == nil {
+			klog.Warningf("Resource %q(%s/%s) already exist in karmada control plane, you can edit PropagationPolicy and OverridePolicy to propagate it\n",
+				gvr, o.Namespace, o.name)
+			return nil
+		}
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to get resource %q(%s/%s) in control plane: %v", gvr, o.Namespace, o.name, err)
+		}
+
+		_, err = controlPlaneDynamicClient.Resource(gvr).Namespace(o.Namespace).Create(context.TODO(), obj, metav1.CreateOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to create resource %q(%s/%s) in control plane: %v", gvr, o.Namespace, o.name, err)
+		}
+		fmt.Printf("ResourceTemplate (%s/%s) is created successfully\n", o.Namespace, o.name)
+	} else {
+		_, err := controlPlaneDynamicClient.Resource(gvr).Get(context.TODO(), o.name, metav1.GetOptions{})
+		if err == nil {
+			klog.Warningf("Resource %q(%s) already exist in karmada control plane, you can edit PropagationPolicy and OverridePolicy to propagate it\n",
+				gvr, o.name)
+			return nil
+		}
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to get resource %q(%s) in control plane: %v", gvr, o.name, err)
+		}
+
+		_, err = controlPlaneDynamicClient.Resource(gvr).Create(context.TODO(), obj, metav1.CreateOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to create resource %q(%s) in control plane: %v", gvr, o.name, err)
+		}
+		fmt.Printf("ResourceTemplate (%s) is created successfully\n", o.name)
+	}
+
+	if o.AutoCreatePolicy {
+		if isNamespaceScoped {
+			policyName, err := o.createPropagationPolicy(karmadaClient, gvr, propagateDeps)
+			if err != nil {
+				return err
+			}
+			fmt.Printf("PropagationPolicy (%s/%s) is created successfully\n", o.Namespace, policyName)
+		} else {
+			policyName, err := o.createClusterPropagationPolicy(karmadaClient, gvr, propagateDeps)
+			if err != nil {
+				return err
+			}
+			fmt.Printf("ClusterPropagationPolicy %s is created successfully\n", policyName)
+		}
+	}
+
+	if isNamespaceScoped {
+		fmt.Printf("Resource %q(%s/%s) is promoted successfully\n", gvr, o.Namespace, o.name)
+	} else {
+		fmt.Printf("Resource %q(%s) is promoted successfully\n", gvr, o.name)
+	}
```

Comment:
The new writing method increases the cyclomatic complexity of functions. It is recommended that the new writing method be consistent with the previous writing method.

Reply:
Makes sense. Thanks! 🙏

Reply:
This makes sense. Thanks! 🙏